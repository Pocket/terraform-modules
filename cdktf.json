{
  "language": "typescript",
  "app": "npm run --silent compile && node dist/src/example.js",
  "terraformProviders": [
    "pagerduty/pagerduty@~> 1.8.0",
    "archive@~> 1.3.0",
    "local@~> 2.1.0"
  ]
}
