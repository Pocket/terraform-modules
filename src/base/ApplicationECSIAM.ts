import { Resource } from 'cdktf';
import { Construct } from 'constructs';
import { iam } from '@cdktf/provider-aws';

export interface ApplicationECSIAMProps {
  prefix: string;
  taskExecutionRolePolicyStatements: iam.DataAwsIamPolicyDocumentStatement[];
  taskRolePolicyStatements: iam.DataAwsIamPolicyDocumentStatement[];
  taskExecutionDefaultAttachmentArn?: string;
  tags?: { [key: string]: string };
}

export class ApplicationECSIAM extends Resource {
  public readonly taskExecutionRoleArn;
  public readonly taskRoleArn;
  public readonly taskRole: iam.IamRole;

  constructor(scope: Construct, name: string, config: ApplicationECSIAMProps) {
    super(scope, name);

    // does anything here need to be in config?
    const dataEcsTaskAssume = new iam.DataAwsIamPolicyDocument(
      this,
      'ecs-task-assume',
      {
        version: '2012-10-17',
        statement: [
          {
            effect: 'Allow',
            actions: ['sts:AssumeRole'],
            principals: [
              {
                identifiers: ['ecs-tasks.amazonaws.com'],
                type: 'Service',
              },
            ],
          },
        ],
      }
    );

    const ecsTaskExecutionRole = new iam.IamRole(this, 'ecs-execution-role', {
      assumeRolePolicy: dataEcsTaskAssume.json,
      name: `${config.prefix}-TaskExecutionRole`,
      tags: config.tags,
    });

    if (config.taskExecutionDefaultAttachmentArn) {
      new iam.IamRolePolicyAttachment(
        this,
        'ecs-task-execution-default-attachment',
        {
          policyArn: config.taskExecutionDefaultAttachmentArn,
          role: ecsTaskExecutionRole.id,
        }
      );
    }

    if (config.taskExecutionRolePolicyStatements.length > 0) {
      const dataEcsTaskExecutionRolePolicy = new iam.DataAwsIamPolicyDocument(
        this,
        'data-ecs-task-execution-role-policy',
        {
          version: '2012-10-17',
          statement: config.taskExecutionRolePolicyStatements,
        }
      );

      const ecsTaskExecutionRolePolicy = new iam.IamPolicy(
        this,
        'ecs-task-execution-role-policy',
        {
          name: `${config.prefix}-TaskExecutionRolePolicy`,
          policy: dataEcsTaskExecutionRolePolicy.json,
        }
      );

      new iam.IamRolePolicyAttachment(
        this,
        'ecs-task-execution-custom-attachment',
        {
          policyArn: ecsTaskExecutionRolePolicy.arn,
          role: ecsTaskExecutionRole.id,
        }
      );
    }

    const ecsTaskRole = new iam.IamRole(this, 'ecs-task-role', {
      assumeRolePolicy: dataEcsTaskAssume.json,
      name: `${config.prefix}-TaskRole`,
      tags: config.tags,
    });

    if (config.taskRolePolicyStatements.length > 0) {
      const dataEcsTaskRolePolicy = new iam.DataAwsIamPolicyDocument(
        this,
        'data-ecs-task-role-policy',
        {
          version: '2012-10-17',
          statement: config.taskRolePolicyStatements,
        }
      );

      const ecsTaskRolePolicy = new iam.IamPolicy(
        this,
        'ecs-task-role-policy',
        {
          name: `${config.prefix}-TaskRolePolicy`,
          policy: dataEcsTaskRolePolicy.json,
        }
      );

      new iam.IamRolePolicyAttachment(this, 'ecs-task-custom-attachment', {
        policyArn: ecsTaskRolePolicy.arn,
        role: ecsTaskRole.id,
      });
    }

    // make arns available to other modules
    this.taskExecutionRoleArn = ecsTaskExecutionRole.arn;
    this.taskRoleArn = ecsTaskRole.arn;
    this.taskRole = ecsTaskRole;
  }
}
