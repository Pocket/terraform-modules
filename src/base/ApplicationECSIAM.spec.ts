import { TerraformStack, Testing } from 'cdktf';
import { ApplicationECSIAM, ApplicationECSIAMProps } from './ApplicationECSIAM';

const BASE_CONFIG: ApplicationECSIAMProps = {
  prefix: 'abides-dev',
  taskExecutionDefaultAttachmentArn: 'someArn',
  taskExecutionRolePolicyStatements: [],
  taskRolePolicyStatements: [],
};

test('renders ECS IAM with minimal config', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  new ApplicationECSIAM(stack, 'testECSService', BASE_CONFIG);

  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('renders ECS IAM with tags', () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, 'test');

  BASE_CONFIG.tags = {
    letsgo: 'bowling',
    donnie: 'throwinrockstonight',
  };

  new ApplicationECSIAM(stack, 'testECSService', BASE_CONFIG);

  expect(Testing.synth(stack)).toMatchSnapshot();
});
