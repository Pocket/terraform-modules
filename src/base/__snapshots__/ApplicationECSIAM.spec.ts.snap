// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ECS IAM with minimal config 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"data\\": {
    \\"aws_iam_policy_document\\": {
      \\"test_testECSService_ecstaskassume_84EDDA06\\": {
        \\"version\\": \\"2012-10-17\\",
        \\"statement\\": [
          {
            \\"effect\\": \\"Allow\\",
            \\"actions\\": [
              \\"sts:AssumeRole\\"
            ],
            \\"principals\\": [
              {
                \\"identifiers\\": [
                  \\"ecs-tasks.amazonaws.com\\"
                ],
                \\"type\\": \\"Service\\"
              }
            ]
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECSService/ecs-task-assume\\",
            \\"uniqueId\\": \\"test_testECSService_ecstaskassume_84EDDA06\\"
          }
        }
      },
      \\"test_testECSService_dataecstaskexecutionrolepolicy_B23D6F8E\\": {
        \\"version\\": \\"2012-10-17\\",
        \\"statement\\": [],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECSService/data-ecs-task-execution-role-policy\\",
            \\"uniqueId\\": \\"test_testECSService_dataecstaskexecutionrolepolicy_B23D6F8E\\"
          }
        }
      }
    }
  },
  \\"resource\\": {
    \\"aws_iam_role\\": {
      \\"test_testECSService_ecsexecutionrole_75EC72C7\\": {
        \\"assume_role_policy\\": \\"\${data.aws_iam_policy_document.test_testECSService_ecstaskassume_84EDDA06.json}\\",
        \\"name\\": \\"abides-dev-TaskExecutionRole\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECSService/ecs-execution-role\\",
            \\"uniqueId\\": \\"test_testECSService_ecsexecutionrole_75EC72C7\\"
          }
        }
      },
      \\"test_testECSService_ecstaskrole_BED63A51\\": {
        \\"assume_role_policy\\": \\"\${data.aws_iam_policy_document.test_testECSService_ecstaskassume_84EDDA06.json}\\",
        \\"name\\": \\"abides-dev-TaskRole\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECSService/ecs-task-role\\",
            \\"uniqueId\\": \\"test_testECSService_ecstaskrole_BED63A51\\"
          }
        }
      }
    },
    \\"aws_iam_role_policy_attachment\\": {
      \\"test_testECSService_ecstaskexecutiondefaultattachment_4EA8C5C9\\": {
        \\"policy_arn\\": \\"someArn\\",
        \\"role\\": \\"\${aws_iam_role.test_testECSService_ecsexecutionrole_75EC72C7.id}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECSService/ecs-task-execution-default-attachment\\",
            \\"uniqueId\\": \\"test_testECSService_ecstaskexecutiondefaultattachment_4EA8C5C9\\"
          }
        }
      },
      \\"test_testECSService_ecstaskexecutioncustomattachment_154A6C8A\\": {
        \\"policy_arn\\": \\"\${aws_iam_policy.test_testECSService_ecstaskexecutionrolepolicy_CAC933F9.arn}\\",
        \\"role\\": \\"\${aws_iam_role.test_testECSService_ecsexecutionrole_75EC72C7.id}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECSService/ecs-task-execution-custom-attachment\\",
            \\"uniqueId\\": \\"test_testECSService_ecstaskexecutioncustomattachment_154A6C8A\\"
          }
        }
      }
    },
    \\"aws_iam_policy\\": {
      \\"test_testECSService_ecstaskexecutionrolepolicy_CAC933F9\\": {
        \\"name\\": \\"abides-dev-TaskExecutionRolePolicy\\",
        \\"policy\\": \\"\${data.aws_iam_policy_document.test_testECSService_dataecstaskexecutionrolepolicy_B23D6F8E.json}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECSService/ecs-task-execution-role-policy\\",
            \\"uniqueId\\": \\"test_testECSService_ecstaskexecutionrolepolicy_CAC933F9\\"
          }
        }
      }
    }
  }
}"
`;

exports[`renders ECS IAM with tags 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"data\\": {
    \\"aws_iam_policy_document\\": {
      \\"test_testECSService_ecstaskassume_84EDDA06\\": {
        \\"version\\": \\"2012-10-17\\",
        \\"statement\\": [
          {
            \\"effect\\": \\"Allow\\",
            \\"actions\\": [
              \\"sts:AssumeRole\\"
            ],
            \\"principals\\": [
              {
                \\"identifiers\\": [
                  \\"ecs-tasks.amazonaws.com\\"
                ],
                \\"type\\": \\"Service\\"
              }
            ]
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECSService/ecs-task-assume\\",
            \\"uniqueId\\": \\"test_testECSService_ecstaskassume_84EDDA06\\"
          }
        }
      },
      \\"test_testECSService_dataecstaskexecutionrolepolicy_B23D6F8E\\": {
        \\"version\\": \\"2012-10-17\\",
        \\"statement\\": [],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECSService/data-ecs-task-execution-role-policy\\",
            \\"uniqueId\\": \\"test_testECSService_dataecstaskexecutionrolepolicy_B23D6F8E\\"
          }
        }
      }
    }
  },
  \\"resource\\": {
    \\"aws_iam_role\\": {
      \\"test_testECSService_ecsexecutionrole_75EC72C7\\": {
        \\"assume_role_policy\\": \\"\${data.aws_iam_policy_document.test_testECSService_ecstaskassume_84EDDA06.json}\\",
        \\"name\\": \\"abides-dev-TaskExecutionRole\\",
        \\"tags\\": {
          \\"letsgo\\": \\"bowling\\",
          \\"donnie\\": \\"throwinrockstonight\\"
        },
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECSService/ecs-execution-role\\",
            \\"uniqueId\\": \\"test_testECSService_ecsexecutionrole_75EC72C7\\"
          }
        }
      },
      \\"test_testECSService_ecstaskrole_BED63A51\\": {
        \\"assume_role_policy\\": \\"\${data.aws_iam_policy_document.test_testECSService_ecstaskassume_84EDDA06.json}\\",
        \\"name\\": \\"abides-dev-TaskRole\\",
        \\"tags\\": {
          \\"letsgo\\": \\"bowling\\",
          \\"donnie\\": \\"throwinrockstonight\\"
        },
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECSService/ecs-task-role\\",
            \\"uniqueId\\": \\"test_testECSService_ecstaskrole_BED63A51\\"
          }
        }
      }
    },
    \\"aws_iam_role_policy_attachment\\": {
      \\"test_testECSService_ecstaskexecutiondefaultattachment_4EA8C5C9\\": {
        \\"policy_arn\\": \\"someArn\\",
        \\"role\\": \\"\${aws_iam_role.test_testECSService_ecsexecutionrole_75EC72C7.id}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECSService/ecs-task-execution-default-attachment\\",
            \\"uniqueId\\": \\"test_testECSService_ecstaskexecutiondefaultattachment_4EA8C5C9\\"
          }
        }
      },
      \\"test_testECSService_ecstaskexecutioncustomattachment_154A6C8A\\": {
        \\"policy_arn\\": \\"\${aws_iam_policy.test_testECSService_ecstaskexecutionrolepolicy_CAC933F9.arn}\\",
        \\"role\\": \\"\${aws_iam_role.test_testECSService_ecsexecutionrole_75EC72C7.id}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECSService/ecs-task-execution-custom-attachment\\",
            \\"uniqueId\\": \\"test_testECSService_ecstaskexecutioncustomattachment_154A6C8A\\"
          }
        }
      }
    },
    \\"aws_iam_policy\\": {
      \\"test_testECSService_ecstaskexecutionrolepolicy_CAC933F9\\": {
        \\"name\\": \\"abides-dev-TaskExecutionRolePolicy\\",
        \\"policy\\": \\"\${data.aws_iam_policy_document.test_testECSService_dataecstaskexecutionrolepolicy_B23D6F8E.json}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testECSService/ecs-task-execution-role-policy\\",
            \\"uniqueId\\": \\"test_testECSService_ecstaskexecutionrolepolicy_CAC933F9\\"
          }
        }
      }
    }
  }
}"
`;
